scylla/src/statement/prepared_statement.rs

```diff
@@ -337,3 +341,132 @@ pub enum PartitionKeyError {
     #[error("Value bytes too long to create partition key, max 65 535 allowed! value.len(): {0}")]
     ValueTooLong(usize),
 }
+
+// TODO: Currently, if there is a null value specified for a partition column,
+// it will be skipped when creating the serialized partition key. We should
+// not create such partition keys in the future, i.e. fail the request or
+// route it to a random replica instead and let the DB reject it. In the
+// meantime, this struct will return some garbage data while it tries to
+// decode the key, but nothing bad like a panic should happen otherwise.
+// The struct is currently only used for printing the partition key, so that's
+// completely fine.
+#[derive(Clone, Copy)]
+pub(crate) struct PartitionKeyDecoder<'pk> {
+    prepared_metadata: &'pk PreparedMetadata,
+    partition_key: &'pk [u8],
+    index: usize,
+}
+
+impl<'pk> PartitionKeyDecoder<'pk> {
+    pub(crate) fn new(prepared_metadata: &'pk PreparedMetadata, partition_key: &'pk [u8]) -> Self {
+        Self {
+            prepared_metadata,
+            partition_key,
+            index: 0,
+        }
+    }
+}
+
+impl<'pk> Iterator for PartitionKeyDecoder<'pk> {
+    type Item = Result<CqlValue, ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.prepared_metadata.pk_indexes.len() {
+            return None;
+        }
+
+        let idx = self.prepared_metadata.pk_indexes[self.index].index as usize;
+        let spec = &self.prepared_metadata.col_specs[idx];
+
+        let cell = if self.prepared_metadata.pk_indexes.len() == 1 {
+            Ok(self.partition_key)
+        } else {
+            self.parse_cell()
+        };
+        self.index += 1;
+        Some(cell.and_then(|mut cell| deser_cql_value(&spec.typ, &mut cell)))
+    }
+}
+
+impl<'pk> PartitionKeyDecoder<'pk> {
+    fn parse_cell(&mut self) -> Result<&'pk [u8], ParseError> {
+        let buf = &mut self.partition_key;
+        let len = buf.read_u16::<BigEndian>()? as usize;
+        if buf.len() < len {
+            return Err(ParseError::IoError(std::io::Error::new(
+                std::io::ErrorKind::UnexpectedEof,
+                "value too short",
+            )));
+        }
+        let col = &buf[..len];
+        *buf = &buf[len..];
+        buf.read_u8()?;
+        Ok(col)
+    }
+}
```

Comment:
Please add some comments for future readers, explaining the format, or at least provide a link to the source describing it.